x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/autoscaling/NodeAvailabilityZoneMapper.java

```diff
@@ -107,30 +108,33 @@ public OptionalInt getNumMlAvailabilityZones() {
     public synchronized void clusterChanged(ClusterChangedEvent event) {
         if (lastDiscoveryNodes == null || event.nodesChanged()) {
             lastDiscoveryNodes = event.state().nodes();
-            buildNodesByAvailabilityZone();
+            updateNodesByAvailabilityZone();
         }
     }
 
-    private synchronized void buildNodesByAvailabilityZone() {
+    private synchronized void updateNodesByAvailabilityZone() {
         if (lastDiscoveryNodes == null) {
             allNodesByAvailabilityZone = Map.of();
             mlNodesByAvailabilityZone = allNodesByAvailabilityZone;
             return;
         }
+        NodesByAvailabilityZone nodesByAvailabilityZone = buildNodesByAvailabilityZone(lastDiscoveryNodes, awarenessAttributes);
+        this.allNodesByAvailabilityZone = nodesByAvailabilityZone.allNodes;
+        this.mlNodesByAvailabilityZone = nodesByAvailabilityZone.mlNodes;
```

Comment:
Are the member variables still immutable?  Before the `Map.copyOf` enforced immutability.  But now are these just hash maps that could be modified?

Reply:
I'm now using `Map.copyOf` when I create a new `NodesByAvailabilityZone` object.